pkg/util/psi.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+const psiLineFormat = "avg10=%f avg60=%f avg300=%f total=%d"
+
+type PSILine struct {
+	Avg10  float64
+	Avg60  float64
+	Avg300 float64
+	Total  uint64
+}
+
+type PSIStats struct {
+	Some *PSILine
+	Full *PSILine
+}
+
+func ReadPSI(pressurePath string) (map[string]float64, error) {
+	defer func() {
+		if e := recover(); e != nil {
+			klog.Warningf("ReadPSI panic: %v, filePath: %s", e, pressurePath)
+		}
+	}()
+	result := map[string]float64{}
+	cpuPressure, err := os.ReadFile(path.Join(pressurePath, "/cpu.pressure"))
+	if err != nil {
+		return result, err
+	}
+	cpuStats, err := parsePSIStats(bytes.NewReader(cpuPressure))
+	if err != nil {
+		return result, err
+	}
+	result["SomeCPUAvg10"] = (*cpuStats.Some).Avg10
+	result["FullMemAvg10"] = 0
```

Comment:
typo -> "FullCPUAvg10".
besides, how to identify whether the `FullCPU` is valued zero or not supported by the kernel (upstream kernel<5.13 does not support FullCPU)

Reply:
use bool field `CPUFullSupported` to identify if `FullCPU` is supported. Notice that if `false` value is set to this field, `FullCPU` will not be reported in prometheus but is inserted into db as zero as golang default nil value.